caps/management/commands/process_scottish_reporting_data.py

```diff
@@ -0,0 +1,428 @@
+                funding_source = funding_source.strip()
+            self.make_data_point(
+                point_type="project",
+                point_data=point_data,
+                sub_type=point_type,
+                lifetime=row["Project lifetime (years)"],
+                cost=row["Capital cost (£)"],
+                funding_source=funding_source,
+                emission_source=row["Primary fuel/emission source saved"],
+                annual_cost=row["Operational cost (£/annum)"],
+                annual_savings=row["Estimated costs savings (£/annum)"],
+                measurement=row[measurement],
+                savings_start=row["First full year of CO2e savings"],
+                comments=row["Comments"],
+            )
+
+        return 1
+
+    def extract_generation_from_sheet(self, df):
+        if len(df.columns) == 3:
+            col = self.get_description_column()
+            titles = self.sheet.iloc[:, col]
+            start_count = 0
+            end_count = 0
+            for index, item in titles.items():
+                if (
+                    type(item) == str
+                    and re.match(self.subsets["generation"]["start_text"], item)
+                    is not None
+                ):
+                    start_count = index + 1
+                if start_count > 0 and index > start_count and pd.isna(item):
+                    end_count = index
+                    break
+
+            if end_count > start_count:
+                row_range = self.sheet.iloc[start_count:end_count, col:]
+                row_range = row_range.dropna(axis="columns", how="all")
+                df = row_range.iloc[1:]
+
+                columns = row_range.iloc[0, 1:].values
+                columns = np.insert(columns, 0, "Technology")
+                df.columns = columns
+
+                data = {}
+                for index, row in df.iterrows():
+                    if row["Technology"] == "Please select from drop down box":
+                        break
+
+                    consumed = 0
+                    exported = 0
+                    for v in row["Total consumed by the body (kWh)"].values:
+                        if v > 0:
+                            consumed = v
+                    for v in row["Total exported (kWh)"].values:
+                        if v > 0:
+                            exported = v
+
+                    self.make_data_point(
+                        point_type="generation",
+                        point_data=row["Technology"],
+                        exported=exported,
+                        consumed=consumed,
+                        comments=row["Comments"],
+                    )
+
+            return 1
+
+    def save_data(self):
+        for name, data in self.data.items():
+            df = pd.DataFrame(data)
+
+            outfile = join(settings.SCOTTISH_DIR, "{}_data.csv".format(name))
+            df.to_csv(open(outfile, "w"), index=False, header=True)
```

Comment:
[it works, so who cares] Any reason to give it the file object rather than just the string?

Reply:
largely because this is the pattern we're using elsewhere and might as well be consistent. I have no actual preference and the pandas docs don't seem to suggest one over the other.